File: src/infrastructure/servers/abstracts/base_server.py
from abc import ABC, abstractmethod
from typing import Any

class BaseServer(ABC):
    def __init__(self):
        _factory = self.create_factory()
        self._server = _factory.create_server()

    @abstractmethod
    def create_factory(self) -> Any:
        pass
    
    def load_server(self):
        return self._server

File: angular-app/src/app/core/infrastructure/servers/abstracts/base-server.ts
import { IServer } from './i-server';
import { BaseServerFactory } from '../factories/abstracts/base-server-factory';

export abstract class BaseServer {
  protected server!: IServer;

  constructor() {
    const factory = this.createFactory();
    this.server = factory.buildServer();
  }

  abstract createFactory(): BaseServerFactory;

  loadServer(): IServer {
    return this.server;
  }
}
